src/Skeleton.hs

```diff
@@ -9,132 +17,274 @@ module Skeleton
+  -- text update event
+  void $ liftGtk $ on entry Gtk.entryActivate $ do
+    putStrLn "entryActivate"
     -- get text
     txt <- Gtk.get entry Gtk.entryText :: IO String
+    triggerStartComputation txt
 
+  computationFinished <- performEventAsync $ ffor startComputation $ \txt done -> do
     -- run "slow" work in thread
-    void $ forkIO $ do
+    void $ liftIO $ forkIO $ do
       putStrLn $ "got text entry event (" <> txt <> "), simulating a worker thread..."
       threadDelay 2000000 -- 2 second delay
       putStrLn "worker thread finished doing work, updating state"
 
-      -- update state after work is done
-      Gtk.postGUISync $
-        modifyIORef stateRef (\(GuiState k _) -> GuiState k (Just txt))
+      done txt
+
+  let stateUpdate = leftmost
+        [ Nothing <$ startComputation
+        , Just <$> computationFinished
+        ]
 
-  pure entry
+  pure (entry, stateUpdate)
 
 
-makeCommitButton :: IORef GuiState -> (GuiState -> IO ()) -> IO Gtk.Button
-makeCommitButton stateRef outputAction = do
-  commitButton <- Gtk.buttonNewWithLabel "commit"
-  Gtk.widgetSetTooltipText commitButton
+makeCommitButton :: _ => m (Gtk.Button, Event t ())
+makeCommitButton = do
+  commitButton <- liftGtk $ Gtk.buttonNewWithLabel "commit"
+  liftGtk $ Gtk.widgetSetTooltipText commitButton
     (Just "SET ME SET ME GO HEAD DO IT COME ON SET ME")
 
-  _ <- on commitButton Gtk.buttonActivated $ do
+  (commit, triggerCommit) <- newTriggerEvent
+  _ <- liftGtk $ on commitButton Gtk.buttonActivated $ do
     putStrLn "commit pressed"
-    readIORef stateRef >>= outputAction
+    triggerCommit ()
+
+  pure (commitButton, commit)
+
+mainWindow :: _ => Event t GuiInputEvent -> (GuiState -> IO ()) -> m Gtk.Window
+mainWindow inputEvent outputAction = do
+  rec stateString <- holdDyn Nothing setStateString
+
+      -- Does this still need to be stateful at all?  I don't think we actually use
+      -- it statefully.  Note that we've extracted the `Just` to here, so now it's
+      -- obvious from examining this one line of code that `stateEvent` will
+      -- transition from `Nothing` to `Just` and then never return to `Nothing`.  We
+      -- could even create a newtype of `Dynamic` that enforces that law if we
+      -- wanted.
+      stateEvent <- holdDyn Nothing $ Just <$> inputEvent
+
+      -- Equivalent to the old `stateRef`
+      let state = GuiState <$> stateEvent <*> stateString
+
+      -- commit button
+      (commitButton, commit) <- makeCommitButton
+      -- I've separated out the logic of what action the commit button triggers
+      -- from its own internal logic here; however, if it makes more sense that
+      -- the commit button do this part itself, this code can easily be pushed
+      -- inside it.  Typically, I find that constraining the effects of each
+      -- function as tightly as possible is a good default approach.
+      --
+      -- Note that if the commit event is simultaneous with a change to the
```

Comment:
Mind adding a commented out version how you would do the post-change version?

Reply:
https://github.com/ghorn/reflex-skeleton/pull/4